syntax = "proto3";

package monitor_rpc;

service MonitorService {
  rpc add_oss(AddOssRequest) returns (AddOssReply) {};
  rpc add_mds(AddMdsRequest) returns (AddMdsReply) {};

  rpc get_oss_cluster(GetOssClusterRequest) returns (GetOssClusterReply) {};

  rpc remove_oss(RemoveOssRequest) returns (RemoveOssReply) {};

  rpc prepare (PrepareRequest) returns (PrepareReply) {};
  rpc accept (AcceptRequest) returns (AcceptReply) {};
  rpc heartbeat (HeartbeatRequest) returns (HeartbeatReply) {};
}

message MdsInfo {
  string name = 1;
  string addr = 2;
}

message OssInfo {
  string name = 1;
  string addr = 2;
}

message GetOssClusterRequest {
  uint64 version = 1;
  string requester = 2;
}

message GetOssClusterReply {
  uint64 version = 1;
  bytes cluster = 2;
  int32 ret_val = 3;
}

message AddOssRequest {
  OssInfo info = 1;
}

message AddOssReply {
  int32 ret_val = 1;
}

message AddMdsRequest {
  MdsInfo info = 1;
}

message AddMdsReply {
  int32 ret_val = 1;
}

message RemoveOssRequest {
  OssInfo info = 1;
}

message RemoveOssReply {
  int32 ret_val = 1;
}



message PrepareRequest {
  uint32 log_index = 1;
  uint64 proposal = 2;
}

message PrepareReply {
  uint64 accepted_proposal = 1;
  bytes accepted_value = 2;
}

message AcceptRequest {
  uint32 log_index = 1;
  uint64 proposal = 2;
  bytes value = 3;
}

message AcceptReply {
  uint64 min_proposal = 1;
}

message HeartbeatRequest {
  string server_name = 1;
}

message HeartbeatReply {

}
